### ISSUE TITLE
Output stream intermittently fails to start 

### ISSUE THREAD

**Issue by jgartman** [google/oboe#1315]
Android version(s): 11
Android device(s): Tab S7, Note 10+
Oboe version: 1.6.0
App name used for testing: OboeTester
 
**Short description**
On certain Android devices I’ve observed an issue where intermittently (maybe 1/100 times just to put a number on it) an output stream will fail to start.

**Steps to reproduce**
This is reproducible in Oboe Tester, here’s a screenshot that shows the settings I’ve been using. 

![Screenshot_20210623-094318_Oboe Tester](https://user-images.githubusercontent.com/12043806/123325414-fd877d80-d505-11eb-93ee-7203f5b4d7eb.jpg)

I typically start the stream, wait a couple of seconds, stop it and then wait a couple more seconds before starting it again.

**Expected behavior**
Stream starts

**Actual behavior**
Stream times out before starting

**Device**
Tab S7:

```
ro.product.brand = samsung
ro.product.manufacturer = samsung
ro.product.model = SM-T970
ro.product.device = gts7xlwifi
ro.product.cpu.abi = arm64-v8a
ro.build.description = gts7xlwifixx-user 11 RP1A.200720.012 T970XXU2BUE2 release-keys
ro.hardware = qcom
ro.hardware.chipname = SM8250
ro.arch = 
| grep aaudio = [aaudio.hw_burst_min_usec]: [2000]
[aaudio.mmap_exclusive_policy]: [1]
[aaudio.mmap_policy]: [1]
```

Also seen on the Note 10+.  I don't have that device at the moment but if needed I can get it to provide more specific info.  

**Any additional context**
Here is a logcat output from an instance where the issue occured:

[logcat.txt](https://github.com/google/oboe/files/6711961/logcat.txt)

Whenever I’ve seen the issue occur it’s always accompanied by this message:
`2021-06-22 18:09:38.905 15875-17093/com.google.sample.oboe.manualtest E/AudioTrack: processAudioBuffer(161): EVENT_MORE_DATA requested 648 bytes but callback returned -1 bytes`

I was looking back through the issue history and I found a couple of other issues that reference that message #396, #810, #569.  I wasn’t sure if this is just a duplicate of those.  Once the stream has failed to start then subsequent attempts to start it always fail.  Closing the stream and re-opening it seems to fix the problem.

We test on other Android 11 Samsung devices such as the S10e and ZFlip and have never encountered the issue on those devices.  For reference here's the device info for the S10e:

```
ro.product.brand = samsung
ro.product.manufacturer = samsung
ro.product.model = SM-G970U
ro.product.device = beyond0q
ro.product.cpu.abi = arm64-v8a
ro.build.description = beyond0qltesq-user 11 RP1A.200720.012 G970USQS5GUE3 release-keys
ro.hardware = qcom
ro.hardware.chipname = SM8150
ro.arch = 
| grep aaudio = [aaudio.hw_burst_min_usec]: [2000]
[aaudio.mmap_exclusive_policy]: [2]
[aaudio.mmap_policy]: [2]
```

**Comment by philburk** [comment]
Thanks for the logcat.txt.  I can see:
> W/AudioStreamLegacy: processCallbackCommon() no data because callback disabled\

That message is from:
http://cs/android/frameworks/av/media/libaaudio/src/legacy/AudioStreamLegacy.cpp?l=96&rcl=9a0fdc393988de23f0852820d8332425752b0066

If mCallbackEnabled is false then it returns SIZE_STOP_CALLBACKS. That may be causing AudioFlinger to permanently kill the stream. So then it may not be able to start again.

mCallbackEnabled is set false in requestStop().
http://cs/android/frameworks/av/media/libaaudio/src/legacy/AudioStreamTrack.cpp?l=381&rcl=9a0fdc393988de23f0852820d8332425752b0066

But sometimes the callback continues after the the call to requestStop(). So there may be a race condition  that occasionally causes a stream to be killed. But this change was added in PI (9.0). ag/3478898 So it seems we would have seen it in Android 10.  But maybe the timing changed in Android 11 for the Samsung device.

**Comment by philburk** [comment]
I hacked in a delay in AudioFlinger and demonstrated that there is a race condition. I have not see this before so I think it is very rare. But it may be more common on certain devices.

The only workaround that I can think of at the moment is to close and reopen the stream after you stop the stream.
Thank you so much for reporting this error.

**Comment by jgartman** [comment]
Thanks for looking into this @philburk , seems like this isn’t an Oboe bug so I really appreciate the info.  Does the legacy aaudio get used exclusively on devices that don’t support MMAP?  I’m trying to get a sense of which devices we might need to be concerned about this issue happening on.

**Comment by philburk** [comment]
The bug is tracked internally with b/192095564 and has been fixed in S.

> Does the legacy aaudio get used exclusively on devices that don’t support MMAP?

Legacy can be used an all Android device. MMAP is sometimes used on some devices, generally higher end devices.

@jgartment - The race exists. The questions is how often does the race cause a failure and on what devices. You mentioned the Tab S7 and Note10+. Do you have data for other devices?

**Comment by jgartman** [comment]
Thanks again Phil, appreciate your responsiveness.  Those are the only two devices that any testers have seen the failure on.  For whatever it's worth, the app has been tested pretty frequently on the following other devices:

Galaxy ZFlip
Galaxy S10e
Galaxy S20 
Tab S6
Galaxy S10+
Pixel 4
Galaxy Note 20
Panasonic FZ-A3
Galaxy S8

**Comment by philburk** [comment]
This was fixed in S in ag/15115530

The only workaround I can think of is to have a watch dog. If the callbacks do not occur within 500 msec then close and reopen the stream. This is not ideal but it should work.
